homeassistant/components/media_source/__init__.py

```diff
@@ -0,0 +1,123 @@
+"""The media_source integration."""
+from datetime import timedelta
+from typing import Optional
+
+import voluptuous as vol
+
+from homeassistant.components import websocket_api
+from homeassistant.components.http.auth import async_sign_path
+from homeassistant.components.media_player.const import ATTR_MEDIA_CONTENT_ID
+from homeassistant.components.media_player.errors import BrowseError
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.integration_platform import (
+    async_process_integration_platforms,
+)
+from homeassistant.loader import bind_hass
+
+from . import local_source, models
+from .const import DOMAIN, URI_SCHEME, URI_SCHEME_REGEX
+from .error import Unresolvable
+
+
+def is_media_source_id(media_content_id: str):
+    """Test if identifier is a media source."""
+    return URI_SCHEME_REGEX.match(media_content_id) is not None
+
+
+def generate_media_source_id(domain: str, identifier: str) -> str:
+    """Generate a media source ID."""
+    uri = f"{URI_SCHEME}{domain or ''}"
```

Comment:
We should always have a domain. That's how we can redirect it to the right source.
```suggestion
    uri = f"{URI_SCHEME}{domain}"
```

Reply:
If we're guarding for when both `domain` and `identifier` is None, maybe we can just return `URI_SCHEME`.

Makes me wonder, with the current code in this function we can just generate `URI_SCHEME`, would that validate with `is_media_source_id` ? 

Makes me wonder if we should have `is_media_source_id` be `media_content_id.startswith(URI_SCHEMA)` and then later when we can't parse it, raise a ValueError ?

Reply:
This is necessary for supporting multiple media source platforms, and it does validate as you can see in [these tests](https://github.com/home-assistant/core/pull/39305/files#diff-4d0b3ce0ca3e18a5394918f0656792bbR14-R17).

`media-source://` - Returns top level "Media Sources" with each domain as a directory to go deeper.
`media-source://domain` - Returns top level for the specified domain.
`media-source://domain/identifier` - Dives deeper into a specific area of a domain.